{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this notebook, we will process the GDSC dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import pubchempy as pcp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "GDSC_link = 'https://cog.sanger.ac.uk/cancerrxgene/GDSC_release8.5/GDSC2_fitted_dose_response_27Oct23.xlsx'\n",
    "if not os.path.exists('../Data/'):\n",
    "    os.mkdir('../Data/')\n",
    "if not os.path.exists('../Data/GDSC2_fitted_dose_response_27Oct23.xlsx'):\n",
    "    !wget -O ../Data/GDSC2_fitted_dose_response_27Oct23.xlsx $GDSC_link"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  DATASET  NLME_RESULT_ID  NLME_CURVE_ID  COSMIC_ID CELL_LINE_NAME  \\\n",
      "0   GDSC2             343       15946310     683667         PFSK-1   \n",
      "1   GDSC2             343       15946548     684052           A673   \n",
      "2   GDSC2             343       15946830     684057            ES5   \n",
      "3   GDSC2             343       15947087     684059            ES7   \n",
      "4   GDSC2             343       15947369     684062          EW-11   \n",
      "\n",
      "  SANGER_MODEL_ID     TCGA_DESC  DRUG_ID     DRUG_NAME PUTATIVE_TARGET  \\\n",
      "0       SIDM01132            MB     1003  Camptothecin            TOP1   \n",
      "1       SIDM00848  UNCLASSIFIED     1003  Camptothecin            TOP1   \n",
      "2       SIDM00263  UNCLASSIFIED     1003  Camptothecin            TOP1   \n",
      "3       SIDM00269  UNCLASSIFIED     1003  Camptothecin            TOP1   \n",
      "4       SIDM00203  UNCLASSIFIED     1003  Camptothecin            TOP1   \n",
      "\n",
      "      PATHWAY_NAME  COMPANY_ID WEBRELEASE  MIN_CONC  MAX_CONC   LN_IC50  \\\n",
      "0  DNA replication        1046          Y    0.0001       0.1 -1.463887   \n",
      "1  DNA replication        1046          Y    0.0001       0.1 -4.869455   \n",
      "2  DNA replication        1046          Y    0.0001       0.1 -3.360586   \n",
      "3  DNA replication        1046          Y    0.0001       0.1 -5.044940   \n",
      "4  DNA replication        1046          Y    0.0001       0.1 -3.741991   \n",
      "\n",
      "        AUC      RMSE   Z_SCORE  \n",
      "0  0.930220  0.089052  0.433123  \n",
      "1  0.614970  0.111351 -1.421100  \n",
      "2  0.791072  0.142855 -0.599569  \n",
      "3  0.592660  0.135539 -1.516647  \n",
      "4  0.734047  0.128059 -0.807232  \n"
     ]
    }
   ],
   "source": [
    "df = pd.read_excel('../Data/GDSC2_fitted_dose_response_27Oct23.xlsx')\n",
    "print(df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Drug Id            Name  Synonyms                            Targets  \\\n",
      "0     1073  5-Fluorouracil      5-FU         Antimetabolite (DNA & RNA)   \n",
      "1     1910          ABT737       NaN  BCL2, BCL-XL,  BCL-W, BCL-B, BFL1   \n",
      "2     1913        AGI-5198   IDH-C35                       IDH1 (R132H)   \n",
      "3     1634        AGI-6780  AK162157                  IDH2 R140Q mutant   \n",
      "4     2045         AMG-319    AMG319                PI3K (beta sparing)   \n",
      "\n",
      "         Target pathway   PubCHEM Datasets  number of cell lines  \\\n",
      "0                 Other      3385    GDSC2                   968   \n",
      "1  Apoptosis regulation  11228183    GDSC2                   957   \n",
      "2            Metabolism  56645356    GDSC2                   958   \n",
      "3            Metabolism  71299339    GDSC2                   893   \n",
      "4   PI3K/MTOR signaling  68947304    GDSC2                   950   \n",
      "\n",
      "  Screening site                                             SMILES  \n",
      "0         SANGER                               C1=C(C(=O)NC(=O)N1)F  \n",
      "1         SANGER  CN(C)CC[C@H](CSC1=CC=CC=C1)NC2=C(C=C(C=C2)S(=O...  \n",
      "2         SANGER  CC1=CC=CC=C1C(C(=O)NC2CCCCC2)N(C3=CC(=CC=C3)F)...  \n",
      "3         SANGER  C1CC1NS(=O)(=O)C2=CC(=C(C=C2)C3=CSC=C3)NC(=O)N...  \n",
      "4         SANGER  C[C@@H](C1=C(N=C2C=C(C=CC2=C1)F)C3=CC=CC=N3)NC...  \n"
     ]
    }
   ],
   "source": [
    "## Process the drug-meta data\n",
    "if not os.path.exists('../Data/drug_meta.csv'):\n",
    "## Download the drug list from https://www.cancerrxgene.org/compounds\n",
    "    drug_meta = pd.read_csv('https://www.cancerrxgene.org/api/compounds?list=all&export=csv')\n",
    "    drug_meta.columns = drug_meta.columns.str.strip()\n",
    "    ## Keep only the GDSC2 drugs\n",
    "    drug_meta = drug_meta[drug_meta['Datasets'] == 'GDSC2']\n",
    "    ## Keep only drugs without missing values in 'PubCHEM'\n",
    "    drug_meta = drug_meta[~drug_meta['PubCHEM'].isna()]\n",
    "    ## Keep the unique drug by the name\n",
    "    drug_meta.drop_duplicates(subset = 'Name', inplace = True, keep='first')\n",
    "    ## Drop the PubCHEM with value 'several'\n",
    "    drug_meta = drug_meta[drug_meta['PubCHEM'] != 'several']\n",
    "    ## Drop the PubCHEM with value 'none'\n",
    "    drug_meta = drug_meta[drug_meta['PubCHEM'] != 'none']\n",
    "    ## Check if there's ',' in the PubCHEM, if yes, keep the first one\n",
    "    drug_meta['PubCHEM'] = drug_meta['PubCHEM'].apply(lambda x: x.split(',')[0])\n",
    "    # Function to get SMILES by PubChem ID\n",
    "    def get_smiles(pubchem_id):\n",
    "        try:\n",
    "            compound = pcp.Compound.from_cid(pubchem_id)\n",
    "            print(f\"SMILES for {pubchem_id} is {compound.isomeric_smiles}\")\n",
    "            return compound.isomeric_smiles\n",
    "        except Exception as e:\n",
    "            return f\"Error: {e}\"\n",
    "\n",
    "    # Add a new column 'SMILES' by applying the function to the 'PubCHEM' column\n",
    "    drug_meta['SMILES'] = drug_meta['PubCHEM'].apply(get_smiles)\n",
    "    drug_meta.to_csv('../Data/drug_meta.csv', index=False)\n",
    "else:\n",
    "    drug_meta = pd.read_csv('../Data/drug_meta.csv')\n",
    "print(drug_meta.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Gene expression as the Features\n",
    "if not os.path.exists('../Data/rnaseq_df.csv'):\n",
    "    exp_df = pd.read_csv(\"https://www.cancerrxgene.org/gdsc1000/GDSC1000_WebResources//Data/preprocessed/Cell_line_RMA_proc_basalExp.txt.zip\", compression = \"zip\", sep = \"\\t\")\n",
    "    exp_df = exp_df.set_index(\"GENE_SYMBOLS\").iloc[:, 1:].T\n",
    "    exp_df.index = exp_df.index.str.extract(\"DATA.([0-9]+)\").to_numpy().squeeze()\n",
    "    exp_df.reset_index(drop=False).groupby(\"index\").first()\n",
    "    ## Make the index from str to int\n",
    "    exp_df.index = exp_df.index.astype(int)\n",
    "    common_cosmic_ids = df['COSMIC_ID'][df['COSMIC_ID'].isin(exp_df.index)].unique()\n",
    "    df = df[df['COSMIC_ID'].isin(common_cosmic_ids)]\n",
    "    exp_df = exp_df.loc[common_cosmic_ids]\n",
    "    exp_df = exp_df.loc[~exp_df.index.duplicated(keep='first')]\n",
    "    exp_df.to_csv('../Data/rnaseq_df.csv')\n",
    "else:\n",
    "    exp_df = pd.read_csv('../Data/rnaseq_df.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Final process of the dataframe\n",
    "df = df[df.COSMIC_ID.isin(exp_df.index)]\n",
    "GDSC2_df = df.merge(drug_meta[['Drug Id', 'PubCHEM']], left_on='DRUG_ID', right_on='Drug Id')[['PubCHEM', 'COSMIC_ID', 'LN_IC50']]\n",
    "GDSC2_df = GDSC2_df[GDSC2_df['COSMIC_ID'].isin(exp_df.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "GDSC2_df = df.merge(drug_meta[['Drug Id', 'PubCHEM']], left_on='DRUG_ID', right_on='Drug Id')[['PubCHEM', 'COSMIC_ID', 'LN_IC50']]\n",
    "GDSC2_df = GDSC2_df[GDSC2_df['COSMIC_ID'].isin(exp_df.index)]\n",
    "## Piviot the GDSC2_df to the table, with PubCHEM as columns, COSMIC_ID as index and LN_IC50 as values\n",
    "GDSC2_df = GDSC2_df.pivot(index='COSMIC_ID', columns='PubCHEM', values='LN_IC50')\n",
    "GDSC2_df.to_csv('../Data/GDSC2_df.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "class GDSCProcessor:\n",
    "    \"\"\"\n",
    "    The GDSCProcessor class automates the download, processing, and preparation of drug response, drug metadata, and gene expression data\n",
    "    from the Genomics of Drug Sensitivity in Cancer (GDSC) project. This class organizes the data into a final dataset that can be used \n",
    "    for further analysis, particularly in drug response predictions.\n",
    "    \n",
    "    Functions:\n",
    "    - download_gdsc_data: Downloads the GDSC2 IC50 data if not already downloaded. 'https://cog.sanger.ac.uk/cancerrxgene/GDSC_release8.5/GDSC2_fitted_dose_response_27Oct23.xlsx'\n",
    "    - process_drug_meta: Downloads and processes drug metadata, retrieves SMILES using PubChem API, and saves results. https://www.cancerrxgene.org/compounds\n",
    "    - process_gene_expression: Downloads and processes gene expression data, and matches it with the drug response data. https://www.cancerrxgene.org/gdsc1000/GDSC1000_WebResources\n",
    "    - process_final_dataframe: Combines drug response data with drug metadata and gene expression data to produce a final dataset.\n",
    "    - run: Executes all the steps to process and produce the final dataset.\n",
    "    \"\"\"\n",
    "    def __init__(self, gdsc_link, drug_meta_link, exp_data_link, data_path = '../Data/', verbose = True):\n",
    "        \"\"\"\n",
    "        Initializes the GDSCProcessor class with URLs for downloading data and setting up the data directory.\n",
    "        \n",
    "        Parameters:\n",
    "        - gdsc_link (str): URL to download the GDSC2 fitted dose response data.\n",
    "        - drug_meta_link (str): URL to download the drug metadata (PubChem, SMILES, etc.).\n",
    "        - exp_data_link (str): URL to download gene expression data.\n",
    "        - data_path (str): Directory to save the processed data (default: '../Data/').\n",
    "        \"\"\"\n",
    "        self.gdsc_link = gdsc_link\n",
    "        self.drug_meta_link = drug_meta_link\n",
    "        self.exp_data_link = exp_data_link\n",
    "        self.data_path = data_path\n",
    "        self.verbose = verbose\n",
    "        if not os.path.exists(self.data_path):\n",
    "            os.mkdir(self.data_path)\n",
    "\n",
    "    def download_gdsc_data(self):\n",
    "        gdsc_file = os.path.join(self.data_path, 'GDSC2_fitted_dose_response_27Oct23.xlsx')\n",
    "        if not os.path.exists(gdsc_file):\n",
    "            print(f\"Downloading GDSC data from {self.gdsc_link}\")\n",
    "            os.system(f\"wget -O {gdsc_file} {self.gdsc_link}\")\n",
    "        self.df = pd.read_excel(gdsc_file)\n",
    "        if self.verbose:\n",
    "            print(self.df.head())\n",
    "    \n",
    "    def process_drug_meta(self):\n",
    "        \"\"\"\n",
    "        Downloads and processes the drug metadata, retrieving SMILES from the PubChem API if necessary.\n",
    "        The processed drug metadata is saved as 'drug_meta.csv'.\n",
    "        If the file already exists, it loads it from the disk.\n",
    "        \"\"\"\n",
    "        drug_meta_file = os.path.join(self.data_path, 'drug_meta.csv')\n",
    "        if not os.path.exists(drug_meta_file):\n",
    "            print('-----------------')\n",
    "            print(f\"Downloading Drug SMILES data\")\n",
    "            self.drug_meta = pd.read_csv(self.drug_meta_link)\n",
    "            self.drug_meta.columns = self.drug_meta.columns.str.strip()\n",
    "            ## Keep only the GDSC2 drugs\n",
    "            self.drug_meta = self.drug_meta[self.drug_meta['Datasets'] == 'GDSC2']\n",
    "            ## Keep only drugs without missing values in 'PubCHEM'\n",
    "            self.drug_meta = self.drug_meta[~self.drug_meta['PubCHEM'].isna()]\n",
    "            ## Keep the unique drug by the name\n",
    "            self.drug_meta.drop_duplicates(subset='Name', inplace=True, keep='first')\n",
    "            ## Drop the PubCHEM with value 'several'\n",
    "            self.drug_meta = self.drug_meta[self.drug_meta['PubCHEM'] != 'several']\n",
    "            ## Drop the PubCHEM with value 'none'\n",
    "            self.drug_meta = self.drug_meta[self.drug_meta['PubCHEM'] != 'none']\n",
    "            self.drug_meta['PubCHEM'] = self.drug_meta['PubCHEM'].apply(lambda x: x.split(',')[0])\n",
    "\n",
    "            def get_smiles(pubchem_id):\n",
    "                try:\n",
    "                    compound = pcp.Compound.from_cid(pubchem_id)\n",
    "                    print(f\"SMILES for {pubchem_id} is {compound.isomeric_smiles}\")\n",
    "                    return compound.isomeric_smiles\n",
    "                except Exception as e:\n",
    "                    return f\"Error: {e}\"\n",
    "\n",
    "            self.drug_meta['SMILES'] = self.drug_meta['PubCHEM'].apply(get_smiles)\n",
    "            self.drug_meta.to_csv(drug_meta_file, index=False)\n",
    "        else:\n",
    "            print(f\"Drug Meta data already exists\")\n",
    "            self.drug_meta = pd.read_csv(drug_meta_file)\n",
    "        if self.verbose:\n",
    "            print(self.drug_meta.head())\n",
    "\n",
    "    \n",
    "    def process_gene_expression(self):\n",
    "        \"\"\"\n",
    "        Downloads and processes gene expression data. It matches the COSMIC IDs from the drug response data \n",
    "        with the gene expression data. The processed gene expression data is saved as 'rnaseq_df.csv'.\n",
    "        \"\"\"\n",
    "        exp_df_file = os.path.join(self.data_path, 'rnaseq_df.csv')\n",
    "        if not os.path.exists(exp_df_file):\n",
    "            print('-----------------')\n",
    "            print(f\"Downloading Gene Expression data\")\n",
    "            self.exp_df = pd.read_csv(self.exp_data_link, compression=\"zip\", sep=\"\\t\")\n",
    "            self.exp_df = self.exp_df.set_index(\"GENE_SYMBOLS\").iloc[:, 1:].T\n",
    "            self.exp_df.index = self.exp_df.index.str.extract(\"DATA.([0-9]+)\").to_numpy().squeeze()\n",
    "            self.exp_df.reset_index(drop=False).groupby(\"index\").first()\n",
    "            self.exp_df.index = self.exp_df.index.astype(int)\n",
    "            common_cosmic_ids = self.df['COSMIC_ID'][self.df['COSMIC_ID'].isin(self.exp_df.index)].unique()\n",
    "            self.df = self.df[self.df['COSMIC_ID'].isin(common_cosmic_ids)]\n",
    "            self.exp_df = self.exp_df.loc[common_cosmic_ids]\n",
    "            self.exp_df = self.exp_df.loc[~self.exp_df.index.duplicated(keep='first')]\n",
    "            self.exp_df.to_csv(exp_df_file)\n",
    "        else:\n",
    "            print(f\"Gene Expression data already exists\")\n",
    "            self.exp_df = pd.read_csv(exp_df_file, index_col=0)\n",
    "\n",
    "    def process_final_dataframe(self):\n",
    "        # Filter the df for matching COSMIC_IDs in exp_df\n",
    "        if not os.path.exists(os.path.join(self.data_path, 'GDSC2_df.csv')):\n",
    "            self.df = self.df[self.df.COSMIC_ID.isin(self.exp_df.index)]\n",
    "\n",
    "            # Merge with drug_meta to get PubCHEM\n",
    "            GDSC2_df = self.df.merge(self.drug_meta[['Drug Id', 'PubCHEM']], left_on='DRUG_ID', right_on='Drug Id')[['PubCHEM', 'COSMIC_ID', 'LN_IC50']]\n",
    "            self.GDSC2_df = GDSC2_df[GDSC2_df['COSMIC_ID'].isin(self.exp_df.index)]\n",
    "\n",
    "            # Pivot the GDSC2_df to a table with PubCHEM as columns, COSMIC_ID as index, and LN_IC50 as values\n",
    "            # GDSC2_df = GDSC2_df.pivot(index='COSMIC_ID', columns='PubCHEM', values='LN_IC50')\n",
    "            GDSC2_df.to_csv(os.path.join(self.data_path, 'GDSC2_df.csv'))\n",
    "            if self.verbose:\n",
    "                print(GDSC2_df.head())\n",
    "        else:\n",
    "            print(f\"Final DataFrame already exists\")\n",
    "            self.GDSC2_df = pd.read_csv(os.path.join(self.data_path, 'GDSC2_df.csv'), index_col=0)\n",
    "            if self.verbose:\n",
    "                print(GDSC2_df.head())\n",
    "\n",
    "    def run(self):\n",
    "        # Step 1: Download GDSC Data\n",
    "        self.download_gdsc_data()\n",
    "\n",
    "        # Step 2: Process Drug Meta Data\n",
    "        self.process_drug_meta()\n",
    "\n",
    "        # Step 3: Process Gene Expression Data\n",
    "        self.process_gene_expression()\n",
    "\n",
    "        # Step 4: Process the final DataFrame\n",
    "        self.process_final_dataframe()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Drug Meta data already exists\n",
      "Gene Expression data already exists\n",
      "Final DataFrame already exists\n"
     ]
    }
   ],
   "source": [
    "# Initialize the class with appropriate URLs\n",
    "processor = GDSCProcessor(\n",
    "    gdsc_link='https://cog.sanger.ac.uk/cancerrxgene/GDSC_release8.5/GDSC2_fitted_dose_response_27Oct23.xlsx',\n",
    "    drug_meta_link='https://www.cancerrxgene.org/api/compounds?list=all&export=csv',\n",
    "    exp_data_link='https://www.cancerrxgene.org/gdsc1000/GDSC1000_WebResources//Data/preprocessed/Cell_line_RMA_proc_basalExp.txt.zip',\n",
    "    verbose = False\n",
    ")\n",
    "\n",
    "# Run the full process\n",
    "processor.run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gnn_tut",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
