{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction to the GDSC Dataset\n",
    "\n",
    "The **Genomics of Drug Sensitivity in Cancer (GDSC)** project is a comprehensive resource aimed at identifying the molecular characteristics of cancer cells that predict sensitivity to anti-cancer drugs. It integrates drug response data from thousands of cell lines with genomic, transcriptomic, and proteomic features to study the impact of genetic variants on drug sensitivity.\n",
    "![GDSC Project Overview](../Figures/GDSC_project.png)\n",
    "*Figure 1: Overview of the GDSC project from the paper \"A Landscape of Pharmacogenomic Interactions in Cancer\" by Iorio et al., published in Cell, Volume 166, Issue 3, pages 740 - 754 (2016).* \n",
    "\n",
    "---\n",
    "\n",
    "## GDSC1\n",
    "\n",
    "**GDSC1** represents the initial phase of the GDSC project, with screenings conducted **before 2015**. This dataset includes:\n",
    "\n",
    "- **970 cancer cell lines**.\n",
    "- **403 drugs** focused primarily on well-established therapeutic agents.\n",
    "\n",
    "## GDSC2\n",
    "\n",
    "**GDSC2** is the expanded phase of the project, with data generated **after 2015**. It includes:\n",
    "\n",
    "- **969 cancer cell lines**, with some overlap but also new cell lines not present in GDSC1.\n",
    "- **297 drugs**, including newer experimental compounds targeting a broader range of cancer pathways.\n",
    "\n",
    "---\n",
    "\n",
    "In this notebook, we will focus on **GDSC2**, as it provides more recent and comprehensive data on drug responses across a wider array of compounds, making it suitable for analyzing modern cancer therapies.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import pubchempy as pcp\n",
    "import sys\n",
    "sys.path.append(os.path.abspath(os.path.join(os.getcwd(), '..')))\n",
    "from utils import *\n",
    "# Initialize the class with appropriate URLs\n",
    "processor = GDSCProcessor(\n",
    "    gdsc_link='https://cog.sanger.ac.uk/cancerrxgene/GDSC_release8.5/GDSC2_fitted_dose_response_27Oct23.xlsx',\n",
    "    drug_meta_link='https://www.cancerrxgene.org/api/compounds?list=all&export=csv',\n",
    "    exp_data_link='https://www.cancerrxgene.org/gdsc1000/GDSC1000_WebResources//Data/preprocessed/Cell_line_RMA_proc_basalExp.txt.zip',\n",
    "    verbose = False,\n",
    "    data_path = '../Data/'\n",
    ")\n",
    "\n",
    "# Run the full process\n",
    "processor.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Drug Response Data\n",
    "\n",
    "The GDSC dataset provides **IC50** (half maximal inhibitory concentration) values for hundreds of drugs across various cancer cell lines. The **IC50** value represents the concentration of a drug required to inhibit a biological process (such as cell growth) by 50%. Lower IC50 values indicate higher sensitivity of the cell line to the drug.\n",
    "\n",
    "### Key Aspects of Drug Response Data:\n",
    "- **IC50 values**: A measure of how effective a drug is at inhibiting cancer cell proliferation.\n",
    "- **Drug ID**: Each drug is assigned a unique ID in the dataset.\n",
    "- **COSMIC_ID**: Each cell line is identified by a unique COSMIC ID, which allows for linking the drug response data to other genomic features of the cell line.\n",
    "\n",
    "The GDSC dataset contains IC50 values for hundreds of drugs tested across over 1,000 cancer cell lines. This information is critical for understanding which cell lines are more sensitive or resistant to particular drugs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 295 drug in GDSC2.\n",
      "There are 969 cancer cell lines in GDSC2.\n"
     ]
    }
   ],
   "source": [
    "## Check how many unique drugs\n",
    "num_drugs = len(processor.df['DRUG_ID'].unique())\n",
    "print(f'There are {num_drugs} drug in GDSC2.') # 295 unique drugs\n",
    "## Check how many unique cell lines\n",
    "num_CCLs = len(processor.df['COSMIC_ID'].unique())\n",
    "print(f'There are {num_CCLs} cancer cell lines in GDSC2.') # 969 unique cell lines"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Cancer Cell Lines \n",
    "\n",
    "Cancer cell lines (CCLs) are used as model systems for studying drug response and genetic features of cancer. The GDSC dataset includes CCLs derived from a variety of cancer types, including breast, lung, colon, and many more. Each CCL has its own set of molecular features, such as gene expression data, mutation status, and copy number variations. In this notebook, we use the RNAseq data, as the omic feature.\n",
    "\n",
    "### Key Aspects of CCL Data:\n",
    "- **COSMIC_ID**: A unique identifier for each CCL, provided by the Catalogue of Somatic Mutations in Cancer (COSMIC) database.\n",
    "- **Gene Expression Profiles**: RNA sequencing data that provides the expression levels of thousands of genes across different CCLs.\n",
    "<!-- - **Mutation Data**: Information on specific mutations found in each CCL, which may contribute to drug sensitivity or resistance. -->\n",
    "\n",
    "By linking drug response data to CCL molecular features, we can investigate the relationships between genetic alterations and drug efficacy.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TSPAN6</th>\n",
       "      <th>TNMD</th>\n",
       "      <th>DPM1</th>\n",
       "      <th>SCYL3</th>\n",
       "      <th>C1orf112</th>\n",
       "      <th>FGR</th>\n",
       "      <th>CFH</th>\n",
       "      <th>FUCA2</th>\n",
       "      <th>GCLC</th>\n",
       "      <th>NFYA</th>\n",
       "      <th>...</th>\n",
       "      <th>LINC00526</th>\n",
       "      <th>PPY2</th>\n",
       "      <th>Unnamed: 17730</th>\n",
       "      <th>Unnamed: 17731</th>\n",
       "      <th>KRT18P55</th>\n",
       "      <th>Unnamed: 17733</th>\n",
       "      <th>POLRMTP1</th>\n",
       "      <th>UBL5P2</th>\n",
       "      <th>TBC1D3P5</th>\n",
       "      <th>Unnamed: 17737</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>683667</th>\n",
       "      <td>7.780713</td>\n",
       "      <td>2.753253</td>\n",
       "      <td>9.960137</td>\n",
       "      <td>4.351073</td>\n",
       "      <td>3.716740</td>\n",
       "      <td>3.222277</td>\n",
       "      <td>8.221606</td>\n",
       "      <td>3.823474</td>\n",
       "      <td>4.756228</td>\n",
       "      <td>5.805642</td>\n",
       "      <td>...</td>\n",
       "      <td>3.347520</td>\n",
       "      <td>3.230713</td>\n",
       "      <td>3.032447</td>\n",
       "      <td>9.040972</td>\n",
       "      <td>3.102091</td>\n",
       "      <td>2.870875</td>\n",
       "      <td>3.169188</td>\n",
       "      <td>9.810430</td>\n",
       "      <td>3.266915</td>\n",
       "      <td>8.452080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>684052</th>\n",
       "      <td>7.301344</td>\n",
       "      <td>2.890533</td>\n",
       "      <td>9.922489</td>\n",
       "      <td>4.125088</td>\n",
       "      <td>3.678987</td>\n",
       "      <td>3.096576</td>\n",
       "      <td>3.588391</td>\n",
       "      <td>4.809305</td>\n",
       "      <td>4.951782</td>\n",
       "      <td>5.089165</td>\n",
       "      <td>...</td>\n",
       "      <td>5.054260</td>\n",
       "      <td>3.003521</td>\n",
       "      <td>2.874737</td>\n",
       "      <td>8.532759</td>\n",
       "      <td>3.068187</td>\n",
       "      <td>2.874065</td>\n",
       "      <td>3.135479</td>\n",
       "      <td>9.073222</td>\n",
       "      <td>3.098364</td>\n",
       "      <td>6.824238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>684057</th>\n",
       "      <td>8.233101</td>\n",
       "      <td>2.824687</td>\n",
       "      <td>10.015884</td>\n",
       "      <td>4.749715</td>\n",
       "      <td>3.839433</td>\n",
       "      <td>3.142754</td>\n",
       "      <td>5.329830</td>\n",
       "      <td>3.272124</td>\n",
       "      <td>5.538055</td>\n",
       "      <td>6.428482</td>\n",
       "      <td>...</td>\n",
       "      <td>6.261573</td>\n",
       "      <td>3.031862</td>\n",
       "      <td>3.370459</td>\n",
       "      <td>8.930821</td>\n",
       "      <td>3.322455</td>\n",
       "      <td>3.083922</td>\n",
       "      <td>2.813440</td>\n",
       "      <td>8.893197</td>\n",
       "      <td>3.266184</td>\n",
       "      <td>8.758289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>684059</th>\n",
       "      <td>8.333466</td>\n",
       "      <td>3.966757</td>\n",
       "      <td>9.793991</td>\n",
       "      <td>3.976923</td>\n",
       "      <td>3.505669</td>\n",
       "      <td>3.079943</td>\n",
       "      <td>3.373640</td>\n",
       "      <td>4.199048</td>\n",
       "      <td>5.794734</td>\n",
       "      <td>5.902391</td>\n",
       "      <td>...</td>\n",
       "      <td>3.885425</td>\n",
       "      <td>2.993918</td>\n",
       "      <td>2.843472</td>\n",
       "      <td>8.246666</td>\n",
       "      <td>3.219777</td>\n",
       "      <td>3.683564</td>\n",
       "      <td>3.033869</td>\n",
       "      <td>8.691401</td>\n",
       "      <td>3.279230</td>\n",
       "      <td>8.236239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>684062</th>\n",
       "      <td>8.391341</td>\n",
       "      <td>2.968360</td>\n",
       "      <td>10.260680</td>\n",
       "      <td>4.295875</td>\n",
       "      <td>4.129471</td>\n",
       "      <td>3.318760</td>\n",
       "      <td>7.103957</td>\n",
       "      <td>3.447994</td>\n",
       "      <td>5.988208</td>\n",
       "      <td>6.257495</td>\n",
       "      <td>...</td>\n",
       "      <td>5.584552</td>\n",
       "      <td>2.959515</td>\n",
       "      <td>2.952987</td>\n",
       "      <td>8.625519</td>\n",
       "      <td>3.056066</td>\n",
       "      <td>3.059551</td>\n",
       "      <td>3.127004</td>\n",
       "      <td>9.396462</td>\n",
       "      <td>3.217885</td>\n",
       "      <td>7.248236</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 17737 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          TSPAN6      TNMD       DPM1     SCYL3  C1orf112       FGR       CFH  \\\n",
       "683667  7.780713  2.753253   9.960137  4.351073  3.716740  3.222277  8.221606   \n",
       "684052  7.301344  2.890533   9.922489  4.125088  3.678987  3.096576  3.588391   \n",
       "684057  8.233101  2.824687  10.015884  4.749715  3.839433  3.142754  5.329830   \n",
       "684059  8.333466  3.966757   9.793991  3.976923  3.505669  3.079943  3.373640   \n",
       "684062  8.391341  2.968360  10.260680  4.295875  4.129471  3.318760  7.103957   \n",
       "\n",
       "           FUCA2      GCLC      NFYA  ...  LINC00526      PPY2  \\\n",
       "683667  3.823474  4.756228  5.805642  ...   3.347520  3.230713   \n",
       "684052  4.809305  4.951782  5.089165  ...   5.054260  3.003521   \n",
       "684057  3.272124  5.538055  6.428482  ...   6.261573  3.031862   \n",
       "684059  4.199048  5.794734  5.902391  ...   3.885425  2.993918   \n",
       "684062  3.447994  5.988208  6.257495  ...   5.584552  2.959515   \n",
       "\n",
       "        Unnamed: 17730  Unnamed: 17731  KRT18P55  Unnamed: 17733  POLRMTP1  \\\n",
       "683667        3.032447        9.040972  3.102091        2.870875  3.169188   \n",
       "684052        2.874737        8.532759  3.068187        2.874065  3.135479   \n",
       "684057        3.370459        8.930821  3.322455        3.083922  2.813440   \n",
       "684059        2.843472        8.246666  3.219777        3.683564  3.033869   \n",
       "684062        2.952987        8.625519  3.056066        3.059551  3.127004   \n",
       "\n",
       "          UBL5P2  TBC1D3P5  Unnamed: 17737  \n",
       "683667  9.810430  3.266915        8.452080  \n",
       "684052  9.073222  3.098364        6.824238  \n",
       "684057  8.893197  3.266184        8.758289  \n",
       "684059  8.691401  3.279230        8.236239  \n",
       "684062  9.396462  3.217885        7.248236  \n",
       "\n",
       "[5 rows x 17737 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "processor.exp_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summary of Cancer Types in GDSC2\n",
    "\n",
    "After processing the dataset, we identified **31 distinct cancer types** (excluding the `UNCLASSIFIED` category). Each cancer type is associated with a unique number of cell lines, which can be used to analyze drug sensitivity patterns.\n",
    "\n",
    "In the GDSC2 dataset, there is one category labeled `UNCLASSIFIED`, representing cell lines that do not fall into the predefined cancer types. For this analysis, we have excluded the `UNCLASSIFIED` category and focused on the main cancer types."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 31 cancer types and one 'UNCLASSIFIED' cancer in GDSC2.\n"
     ]
    }
   ],
   "source": [
    "cancer_df = processor.df.groupby('TCGA_DESC')['COSMIC_ID'].nunique().reset_index()\n",
    "cancer_df.set_index('TCGA_DESC', inplace=True)\n",
    "cancer_df.columns = ['Number of Unique Cell Lines']\n",
    "cancer_df.sort_values('Number of Unique Cell Lines', ascending=False, inplace=True)\n",
    "## Drop UNCLASSIFIED and nan in 'TCGA_DESC'\n",
    "cancer_df = cancer_df[cancer_df.index != 'UNCLASSIFIED'].dropna()\n",
    "print(f'There are {len(cancer_df)} cancer types and one \\'UNCLASSIFIED\\' cancer in GDSC2.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Top 10 Cancer Types by Number of Unique Cell Lines\n",
    "\n",
    "The table below shows the **top 10 cancer types** with the highest number of unique cell lines in the GDSC2 dataset:\n",
    "\n",
    "| Cancer Type (TCGA_DESC) | Number of Unique Cell Lines |\n",
    "|-------------------------|-----------------------------|\n",
    "| LUAD (Lung Adenocarcinoma)     | 62                          |\n",
    "| SCLC (Small Cell Lung Cancer)  | 59                          |\n",
    "| SKCM (Skin Cutaneous Melanoma) | 54                          |\n",
    "| BRCA (Breast Cancer)           | 51                          |\n",
    "| COREAD (Colorectal Cancer)     | 46                          |\n",
    "| HNSC (Head and Neck Cancer)    | 39                          |\n",
    "| ESCA (Esophageal Cancer)       | 35                          |\n",
    "| DLBC (Diffuse Large B-Cell Lymphoma) | 34                   |\n",
    "| GBM (Glioblastoma Multiforme)  | 34                          |\n",
    "| OV (Ovarian Cancer)            | 34                          |\n",
    "\n",
    "These cancer types represent the most frequently occurring cancer cell lines in the dataset and provide a broad range of data for drug sensitivity analysis.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           Number of Unique Cell Lines\n",
      "TCGA_DESC                             \n",
      "LUAD                                62\n",
      "SCLC                                59\n",
      "SKCM                                54\n",
      "BRCA                                51\n",
      "COREAD                              46\n"
     ]
    }
   ],
   "source": [
    "print(cancer_df.head(5))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Drugs\n",
    "\n",
    "In addition to drug response data, the GDSC dataset includes information about the drugs themselves. Each drug is characterized by various chemical properties that can be used to predict its effectiveness in inhibiting cancer cell growth.\n",
    "\n",
    "### Key Aspects of Drug Features:\n",
    "- **PubChem ID**: A unique identifier for each drug in the PubChem database.\n",
    "- **SMILES**: A textual representation of the molecular structure of the drug, which can be used to calculate molecular descriptors for quantitative structure-activity relationship (QSAR) modeling.\n",
    "- **Pathway Information**: Many drugs target specific biological pathways (e.g., PI3K/AKT/mTOR signaling) that are dysregulated in cancer cells. Understanding these pathways helps identify which drugs might be effective against particular cell lines.\n",
    "<!-- - **Molecular Descriptors**: Features such as molecular weight, LogP (partition coefficient), and hydrogen bond donors/acceptors provide additional information about the chemical properties of each drug. These descriptors can be used to predict drug efficacy. -->\n",
    "\n",
    "By integrating drug features with cell line molecular data, we can build predictive models for drug response, identify potential biomarkers of sensitivity, and uncover novel therapeutic strategies for cancer treatment."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Drug Target Pathways in the GDSC2 Dataset\n",
    "\n",
    "The **Genomics of Drug Sensitivity in Cancer (GDSC)** dataset not only provides drug response data but also links each drug to the biological pathways they target. This helps in understanding the mechanisms through which drugs act on different cancer cell lines and allows for the identification of common pathways targeted by multiple drugs.\n",
    "\n",
    "## Summary of Drug Target Pathways in GDSC2\n",
    "\n",
    "In the GDSC2 dataset, drugs are categorized based on the pathways they target. Some drugs may target multiple pathways, but each drug is uniquely associated with its primary pathway in the dataset. These pathways represent key biological processes often dysregulated in cancer, such as **PI3K/MTOR signaling**, **DNA replication**, and **ERK MAPK signaling**.\n",
    "\n",
    "### Key Drug Target Pathways\n",
    "\n",
    "After processing the dataset, we identified **22 distinct pathways** associated with drug targets, excluding the `Unclassified` and `Other` categories. These pathways play critical roles in cancer development and progression, and targeting these pathways can lead to potential therapeutic interventions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 22 pathways in GDSC2, one 'Unclassified', and  one 'Other' drug target pathway.\n"
     ]
    }
   ],
   "source": [
    "processor.df['PATHWAY_NAME']\n",
    "\n",
    "drug_df = processor.df.groupby('PATHWAY_NAME')['DRUG_ID'].nunique().reset_index()\n",
    "drug_df.set_index('PATHWAY_NAME', inplace=True)\n",
    "drug_df.columns = ['Number of Unique Drugs']\n",
    "drug_df.sort_values('Number of Unique Drugs', ascending=False, inplace=True)\n",
    "drug_df = drug_df[drug_df.index != 'Unclassified'].dropna()\n",
    "drug_df = drug_df[drug_df.index != 'Other'].dropna()\n",
    "print(f'There are {len(drug_df)} pathways in GDSC2, one \\'Unclassified\\', and  one \\'Other\\' drug target pathway.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "### Top 10 Drug Target Pathways by Number of Unique Drugs\n",
    "\n",
    "The table below shows the **top 10 drug target pathways** with the highest number of unique drugs in the GDSC2 dataset:\n",
    "\n",
    "| Pathway Name                  | Number of Unique Drugs |\n",
    "|-------------------------------|------------------------|\n",
    "| PI3K/MTOR signaling            | 27                     |\n",
    "| Other, kinases                 | 22                     |\n",
    "| DNA replication                | 21                     |\n",
    "| ERK MAPK signaling             | 15                     |\n",
    "| Chromatin histone methylation  | 13                     |\n",
    "| Apoptosis regulation           | 13                     |\n",
    "| Cell cycle                     | 13                     |\n",
    "| Genome integrity               | 13                     |\n",
    "| RTK signaling                  | 12                     |\n",
    "| Chromatin other                | 10                     |\n",
    "\n",
    "These pathways represent the most frequently targeted mechanisms by drugs in the dataset. **PI3K/MTOR signaling** is the most targeted pathway, with 27 unique drugs, followed by **DNA replication** and **ERK MAPK signaling**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Number of Unique Drugs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PATHWAY_NAME</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PI3K/MTOR signaling</th>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Other, kinases</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNA replication</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERK MAPK signaling</th>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Chromatin histone methylation</th>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               Number of Unique Drugs\n",
       "PATHWAY_NAME                                         \n",
       "PI3K/MTOR signaling                                27\n",
       "Other, kinases                                     22\n",
       "DNA replication                                    21\n",
       "ERK MAPK signaling                                 15\n",
       "Chromatin histone methylation                      13"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drug_df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gnn_tut",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
